internal/app/api/v1/testsource.go

```diff
@@ -131,6 +133,73 @@ func (s TestkubeAPI) DeleteTestSourcesHandler() fiber.Handler {
 	}
 }
 
+func (s TestkubeAPI) ProcessTestSourceBatchHandler() fiber.Handler {
+	return func(c *fiber.Ctx) error {
+		var request []testkube.TestSourceUpsertRequest
+		err := c.BodyParser(&request)
+		if err != nil {
+			return s.Error(c, http.StatusBadRequest, err)
+		}
+
+		testSourceBatch := make(map[string]testkube.TestSourceUpsertRequest, len(request))
+		for _, item := range request {
+			if _, ok := testSourceBatch[item.Name]; !ok {
+				return s.Error(c, http.StatusBadRequest, fmt.Errorf("test source with duplicted id/name %s", item.Name))
+			}
+
+			testSourceBatch[item.Name] = item
+		}
+
+		list, err := s.TestSourcesClient.List("")
+		if err != nil {
+			return s.Error(c, http.StatusBadRequest, err)
+		}
+
+		testSourceMap := make(map[string]testsourcev1.TestSource, len(list.Items))
+		for _, item := range list.Items {
+			testSourceMap[item.Name] = item
+		}
+
+		var result testkube.TestSourceBatchResult
+		for name, item := range testSourceBatch {
+			testSource := testsourcesmapper.MapAPIToCRD(item)
+			if existed, ok := testSourceMap[name]; !ok {
+				testSource.Namespace = s.Namespace
+
+				created, err := s.TestSourcesClient.Create(&testSource, testsources.Option{Secrets: getTestSourceSecretsData(&item)})
+				if err != nil {
+					return s.Error(c, http.StatusBadRequest, err)
+				}
+
+				result.Created = append(result.Created, created.Name)
+			} else {
+				existed.Spec = testSource.Spec
+				existed.Labels = item.Labels
+
+				updated, err := s.TestSourcesClient.Update(&existed, testsources.Option{Secrets: getTestSourceSecretsData(&item)})
+				if err != nil {
+					return s.Error(c, http.StatusBadGateway, err)
+				}
+
+				result.Updated = append(result.Updated, updated.Name)
+			}
+		}
+
+		for name := range testSourceMap {
+			if _, ok := testSourceBatch[name]; !ok {
+				err := s.TestSourcesClient.Delete(name)
+				if err != nil {
+					return s.Error(c, http.StatusBadRequest, err)
```

Comment:
Good question. In test suite we have a field to fail fast or not. There we should have defaault behaivour, I prefer to fail fast in order to avoid side affects. And on practice, it will fail, if something is wrong with k8s cluster access, mean, other calls will fail as well